larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewAmbiguousDeltaRayTool.cc

```diff
@@ -0,0 +1,108 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewAmbiguousDeltaRayTool.cc
+ *
+ *  @brief  Implementation of the two view amgiuous delta ray tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewAmbiguousDeltaRayTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewAmbiguousDeltaRayTool::TwoViewAmbiguousDeltaRayTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewAmbiguousDeltaRayTool::Run(TwoViewDeltaRayMatchingAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+       std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    this->ExamineConnectedElements(pAlgorithm, overlapMatrix);
+
+    // ATTN: Prevent tensor tool loop running again
+    return false;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewAmbiguousDeltaRayTool::ExamineConnectedElements(TwoViewDeltaRayMatchingAlgorithm *const pAlgorithm, MatrixType &overlapMatrix) const
+{
+    bool particleCreated(true);
+
+    while (particleCreated)
+    {
+        particleCreated = false;
+
+        ClusterVector sortedKeyClusters;
+        overlapMatrix.GetSortedKeyClusters(sortedKeyClusters);
+
+        ClusterSet usedKeyClusters;
+
+        for (const Cluster *const pKeyCluster : sortedKeyClusters)
+        {
+            if (usedKeyClusters.count(pKeyCluster))
+                continue;
+
+            MatrixType::ElementList elementList;
+            overlapMatrix.GetConnectedElements(pKeyCluster, true, elementList);
+
+            for (const MatrixType::Element &element : elementList)
+                usedKeyClusters.insert(element.GetCluster1());
+
+            if (this->PickOutGoodMatches(pAlgorithm, elementList))
+            {
+                particleCreated = true;
+                break;
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewAmbiguousDeltaRayTool::PickOutGoodMatches(TwoViewDeltaRayMatchingAlgorithm *const pAlgorithm, const MatrixType::ElementList &elementList) const
+{
+    unsigned int highestHitCount(0);
+    float bestChiSquared(std::numeric_limits<float>::max());
+    MatrixType::Element bestElement(nullptr, nullptr, TwoViewDeltaRayOverlapResult(TwoViewXOverlap(0.f, 0.f, 0.f, 0.f), PfoList(), nullptr, ClusterList(), 0.f));
```

Comment:
Rather than creating a dummy object to hold the best element, you could opt for a best index variable, initialised to -1, (you'd need to either maintain a loop counter or adopt the classic for loop style over the list) and then assuming the a best index is found, just use that to access the appropriate element of the list when creating the pfo